### PR TITLE
[AutoDiff upstream] Upstream `@derivative` attribute type-checking.

### PR DESCRIPTION
**Pull request by dan-zheng**
The `@derivative` attribute registers a function as a derivative of another
function-like declaration: a `func`, `init`, `subscript`, or `var` computed
property declaration.

The `@derivative` attribute also has an optional `wrt:` clause specifying the
parameters that are differentiated "with respect to", i.e. the differentiation
parameters. The differentiation parameters must conform to the `Differentiable`
protocol.

If the `wrt:` clause is unspecified, the differentiation parameters are inferred
to be all parameters that conform to `Differentiable`.

`@derivative` attribute type-checking verifies that the type of the derivative
function declaration is consistent with the type of the referenced original
declaration and the differentiation parameters.

Resolves TF-829.

---

Examples:
```swift
func sin(_ x: Float) -> Float { ... }

@derivative(of: sin)
func derivativeSin(_ x: Float) -> (value: Float, differential: (Float) -> Float) {
    (sin(x), { dx in dx * cos(x) })
}

public protocol ElementaryFunctions {
    static func log(_ x: Self) -> Self
    ...
}

public extension ElementaryFunctions
where Self: Differentiable & FloatingPoint, Self == Self.TangentVector {
    @inlinable
    @derivative(of: log)
    static func derivativeLog(_ x: Self) -> (value: Self, differential: (Self) -> Self) {
        (log(x), { dx in (1 / x) * dx })
    }
}
```

The [differentiable programming manifesto](https://github.com/apple/swift/blob/master/docs/DifferentiableProgramming.md#make-a-function-differentiable-using-derivative-or-transpose) has more information and examples.

---

Note: this code is upstreamed from `tensorflow` branch. Some functions are intentionally not `static` (fileprivate) because they have users in other files on `tensorflow` branch.

Known issues are referenced in comments:
* TF-874: Simplify type calculation helpers.
* TF-1042: Remove `DerivativeAttrs` cache from `ASTContext`.

We would like to upstream code and continue development on `master`, if that's acceptable!

cc @rxwei @marcrasi @bgogul

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test

File: lib/AST/AutoDiff.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
+//===--------- AutoDiff.cpp - Swift Differentiable Programming ------------===//
```
Comment (lattner) [inline lib/AST/AutoDiff.cpp]: nit: please use the standard header guard, which left aligns the filename, and please change the thing after it to be descriptive, not a marketing thing :-)

**Review by lattner — APPROVED** [review]
Nice, I love the testcase, great job.

**Review by rxwei — APPROVED** [review]

File: lib/AST/AutoDiff.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/AST/AutoDiff.cpp]: Done in https://github.com/apple/swift/pull/28738/commits/2ce6b05d6efcbd6a1c9a1c95d2bd6597fac57252.

I noticed that `IndexSubset.{h,cpp}` also use inconsistent headers, so I fixed those in https://github.com/apple/swift/pull/28757.

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by dan-zheng** [comment]
@swift-ci Please test macOS

**Comment by dan-zheng** [comment]
Merging to unblock progress on upstreaming AutoDiff. Happy to address additional feedback later!

**Comment by compnerd** [comment]
This broke the VS2019 build:
https://compnerd.visualstudio.com/windows-swift/_build/results?buildId=18893&view=logs&j=b9e62f99-1a98-5ed7-01d2-f4794231ed79&t=9d1c7fdf-e7ea-5e83-bbfd-5e0708569c31&l=3865

**Pull request merged**